HW2/P4/driver.py

```diff
@@ -15,15 +15,44 @@
 from timer import Timer
 import threading
 
+# Helper function to make sure we don't wait
+# forever on the first iteration or break on the last one.
+def bounds_check(j, num_threads):
+    return j >= 0 and j < num_threads
+
+def median_worker(tmpa, tmpb, j, events, num_threads, iters):
+    for i in range(iters):
+        # Wait on all the necessary conditions in order to execute (extra credit).
+        if bounds_check(j - 1, num_threads) and i - 1 >= 0:
+            events[i - 1][j - 1].wait()
+        if bounds_check(j, num_threads) and i - 1 >= 0:
+            events[i - 1][j].wait()
+        if bounds_check(j + 1, num_threads) and i - 1 >= 0:
+            events[i - 1][j + 1].wait()
+        filtering.median_3x3(tmpa, tmpb, j, num_threads)
```

Comment:
I did this using exactly the same way by creating a 2-D array of events to record the thread status. While the 'bounds_check()' is necessary, I don't think it is a good idea to check in every iteration, simply changing the starting point could be a better solution that avoids redundant checks.